homeassistant/components/foscam/camera.py

```diff
@@ -28,20 +60,70 @@
     }
 )
 
+SERVICE_PTZ_SCHEMA = vol.Schema(
+    {
+        vol.Required(ATTR_ENTITY_ID): cv.entity_ids,
+        vol.Required(ATTR_MOVEMENT): vol.In(
+            [
+                DIR_UP,
+                DIR_DOWN,
+                DIR_LEFT,
+                DIR_RIGHT,
+                DIR_TOPLEFT,
+                DIR_TOPRIGHT,
+                DIR_BOTTOMLEFT,
+                DIR_BOTTOMRIGHT,
+            ]
+        ),
+        vol.Optional(ATTR_TRAVELTIME, default=DEFAULT_TRAVELTIME): cv.small_float,
+    }
+)
+
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+async def async_setup_platform(hass, config, async_add_entities, discovery_info=None):
     """Set up a Foscam IP Camera."""
-    add_entities([FoscamCam(config)])
+    _LOGGER.debug("Setting up the Foscam camera platform")
+
+    async def async_handle_ptz(service):
+        """Handle PTZ service call."""
+        movement = service.data.get(ATTR_MOVEMENT)
+        travel_time = service.data.get(ATTR_TRAVELTIME)
+        all_cameras = hass.data[FOSCAM_DATA][ENTITIES]
+        entity_ids = await async_extract_entity_ids(hass, service)
+        target_cameras = []
+
+        _LOGGER.debug("Moving '%s' camera(s): %s", movement, entity_ids)
+
+        if not entity_ids:
+            target_cameras = all_cameras
+        else:
+            target_cameras = [
+                camera for camera in all_cameras if camera.entity_id in entity_ids
+            ]
+
+        for camera in target_cameras:
+            await camera.async_perform_ptz(movement, travel_time)
+
+    hass.services.async_register(
+        DOMAIN, SERVICE_PTZ, async_handle_ptz, schema=SERVICE_PTZ_SCHEMA
+    )
+
+    _LOGGER.debug("Constructing the FoscamCam")
+    async_add_entities([FoscamCam(hass, config)])
 
 
 class FoscamCam(Camera):
     """An implementation of a Foscam IP camera."""
 
-    def __init__(self, device_info):
+    def __init__(self, hass, device_info):
```

Comment:
Don't pass in `hass`. It will be set on the entity when it has been added to home assistant.